tests/test_class_sh_shared_ptr_copy_move.cpp

```diff
@@ -0,0 +1,62 @@
+#include "pybind11_tests.h"
+
+#include <pybind11/smart_holder.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace pybind11_tests {
+namespace class_sh_shared_ptr_copy_move {
+
+template<int SerNo>
+struct Foo {
+    std::string mtxt;
+    Foo() : mtxt("DefaultConstructor") {}
+    Foo(const std::string &mtxt_) : mtxt(mtxt_) {}
+    Foo(const Foo &other) { mtxt = other.mtxt + "_CpCtor"; }
+    Foo(Foo &&other) { mtxt = other.mtxt + "_MvCtor"; }
+};
+
+using FooAVL = Foo<0>;
+using FooDEF = Foo<1>;
+
+} // namespace class_sh_shared_ptr_copy_move
+} // namespace pybind11_tests
+
+PYBIND11_SMART_HOLDER_TYPE_CASTERS(pybind11_tests::class_sh_shared_ptr_copy_move::FooAVL)
+
+namespace pybind11_tests {
+namespace class_sh_shared_ptr_copy_move {
+
+TEST_SUBMODULE(class_sh_shared_ptr_copy_move, m) {
+    namespace py = pybind11;
+
+    py::class_<FooAVL, PYBIND11_SH_AVL(FooAVL)>(m, "FooAVL");
+    py::class_<FooDEF, PYBIND11_SH_DEF(FooDEF)>(m, "FooDEF");
```

Comment:
Suggested change:
```
    py::class_<FooShPtr, std::shared_ptr<FooAVL>>(m, "FooShPtr");
    py::classh<FooSmHld>(m, "FooSmHld");
```

What I had in mind is unconditionally comparing the `std::shared_ptr` vs `py::smart_holder` behavior.
Feel free to pick different variable names.

You will also need the `PYBIND11_TYPE_CASTER_BASE_HOLDER` macro for `FooShPtr`.

To make that easier to read, in more recent code I'm using the anonymous `namespace` instead of `pybind11_tests::class_sh_shared_ptr_copy_move`. I think it looks much better and is just as safe.

Reply:
Ah, ok. Should be updated accordingly now.